python-package/xgboost/core.py

```diff
@@ -138,27 +138,42 @@ def c_array(ctype, values):
     return (ctype * len(values))(*values)
 
 
-def _maybe_from_pandas(data, feature_names, feature_types):
+def _maybe_from_pandas(data, label, feature_names, feature_types):
     """ Extract internal data from pd.DataFrame """
     try:
         import pandas as pd
     except ImportError:
-        return data, feature_names, feature_types
+        return data, label, feature_names, feature_types
 
     if not isinstance(data, pd.DataFrame):
-        return data, feature_names, feature_types
-
-    dtypes = data.dtypes
-    if not all(dtype.name in ('int64', 'float64', 'bool') for dtype in dtypes):
-        raise ValueError('DataFrame.dtypes must be int, float or bool')
+        return data, label, feature_names, feature_types
+
+    mapper = {'int8': 'int', 'int16': 'int', 'int32': 'int', 'int64': 'int',
+              'uint8': 'int', 'uint16': 'int', 'uint32': 'int', 'uint64': 'int',
+              'float16': 'float', 'float32': 'float', 'float64': 'float',
+              'bool': 'i'}
+
+    data_dtypes = data.dtypes
+    if not all(dtype.name in (mapper.keys()) for dtype in data_dtypes):
+        raise ValueError('DataFrame.dtypes for data must be int, float or bool')
+
+    if label is not None:
+        if isinstance(label, pd.DataFrame):
```

Comment:
@sinhrks 
- I didn't put the extra check there, because when you use a multi dimensional ```DataFrame```, the function ```set_label()```  generates the same error as when you use a multi dimensional ```numpy.ndarray```, namely ```TypeError: only length-1 arrays can be converted to Python scalars```. The check is just to make sure all values get converted to ```float```.
- ```Series``` already worked the same as a ```list``` or ```numpy.ndarray```. All get converted by the function ```set_label()```.
- If you want to check for multi dimensional data, I think it is better to alter the ```__init__``` code inside the ```DMatrix``` class itself, just before ```self.set_label(label)```.